### ISSUE TITLE
Check for launch configuration limits in CUDA backend.

### ISSUE THREAD

**Issue by pavanky** [arrayfire/arrayfire#133]
Verify against [`cudaDeviceProp`](http://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp) instead of a hard coded value like 65535.

**Comment by pavanky** [comment]
@umar456 This is only going to be a limitation for pre-kepler cards. Do we need to worry about the issue or leave it as a known issue ?

**Comment by umar456** [comment]
We should support everything CUDA supports. At least things that are not deprecated in the new versions.

**Comment by pavanky** [comment]
hmm  but it only affects Fermi GPUs in weird configurations (eg: very tall). The architecture is 5 years old at this point and adding support around the limitation requires quite a bit of change in code.

**Comment by pavanky** [comment]
I stand corrected. The limit for y and z dimensions is still 65535 and we hit this limit for certain configurations. We should look into fixing this properly.

**Comment by pavanky** [comment]
@shehzan10 do you want to take a shot at figuring out how to deal with this ?

**Comment by vakopian** [comment]
I started hitting this bug recently too. Other reports have indicated a large second dimension being the culprit, but I think I have an example without a large second dimension:
```
    int sz = 8000;
    array A = randu(1000000, 3); // 1000000x3
    array idx = tile(randu(sz), 1, 3, 300); // 8000x3x300
    array B = A(idx, span); // 7200000x3
```
This produces
```
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
In function void cuda::evalNodes(std::vector<cuda::Param<T> >&, std::vector<cuda::JIT::Node*>) [with T = unsigned int]
In file src/backend/cuda/jit.cpp:658
CU Error (1)

In function af::array::array_proxy::operator af::array()
In file src/api/cpp/array.cpp:705
Signal: SIGABRT (Aborted)
```
It does work if I reduce `sz` to `6000`.
af_info:
```
ArrayFire v3.4.2 (CUDA, 64-bit Linux, build 801f40f)
Platform: CUDA Toolkit 8, Driver: 367.48
[0] Quadro K2200, 4042 MB, CUDA Compute 5.0
```

**Comment by InBetweenNames** [comment]
What would it take to get this issue fixed in ArrayFire?  I'm able to work around it by only processing up to 65535 elements at once, but it's a bit of a hassle to do and it affects code clarity.

**Comment by pavanky** [comment]
Fixed in v3.5
